lib/Fencer/Rules.hs

```diff
@@ -100,6 +118,26 @@ loadRulesFromDirectory
         dirs <- filterM isDirectory other
         (files ++) <$> concatMapM listAllFiles dirs
 
+-- | Perform validation checks to make sure the behavior matches that
+-- of @lyft/ratelimit@.
+validatePotentialDomains
+  :: [Either LoadRulesError (Maybe DomainDefinition)]
+  -> Either [LoadRulesError] [DomainDefinition]
+validatePotentialDomains res = case partitionEithers res of
+  (errs@(_:_), _    ) -> Left errs
+  ([]        , mRules) -> do
+    -- check if there are any duplicate domains
+    let
+      rules = catMaybes mRules
+      groupedRules :: [NonEmpty DomainDefinition] = NE.groupBy
+        ((==) `on` (unDomainId . domainDefinitionId))
+        (NE.fromList $ sortOn domainDefinitionId rules)
+    if (length @[] rules /= length @[] groupedRules)
```

Comment:
Do you need those `@[]`? Or are they just for documentation purposes?

Reply:
The `@[]`s are required by HLint.

I added a rule to forbid unqualified `length` and `null` because otherwise you don't get a compiler error when a list changes to e.g. a Maybe list during a refactoring.

Reply:
Never had any such problem. And `Maybe` is a 1-element container, it's perfectly fine to refactor an n-element container into a 1-element one, but keep `length`, `null`, etc. It looks to me that the semantics will be preserved, so I do not see why bother with specifying calls to `length` and `null`. But not a big deal.

Reply:
It's fine to refactor `[a]` into `Maybe a`, but when you refactor `[a]` into `Maybe [a]` or `(x, [a])`, the semantics are (often) no longer preserved and the silent breakage is more likely.

> Never had any such problem.

Not claiming it's a particularly common problem. I did stumble upon it, though.

Reply:
Good point.